keps/sig-scheduling/20180409-scheduling-framework.md

```diff
@@ -264,6 +268,31 @@ state, it will either trigger [Un-reserve](#un-reserve) plugins (on failure) or
 
 *Note: This concept used to be referred to as "assume".*
 
+### Pre-Preempt
+
+This is an informational extension point. Plugins will be called with a pod which was unscheduled and eligible to preemption. A plugin may use this data to update internal state or to generate logs/metrics.
+
+### Preempt
+
+The plugin provides an extension point for selecting the victim pods for preemption.  Only one preempt plugin may be enabled at a time.  
```

Comment:
What if we make this endpoint more similar to `Filter`. i.e. there can be several, one runs after the other. They all have to pass. Then kube-scheduler can implement a default one, based on priority classes, which runs first.

Reply:
This proposal makes current preemption victim logic just hooks.  As you might know, current victim selection logic 

- select victim candidates for _each node_: I think this phase can accept multiple plugins.
- choose the victim node finally:  I think this phase would be hard to accept multiple plugins.

Thus,  I added this restriction here.

Reply:
> But DryRun is not exactly Filter because each DryRun can generate its own []Victims - which are unioned rather than intersected. (technically they cannot be intersected either)

Yeah,  this looks generic enough for me :-)